Change interoperability genesis asset schema (#256)
LiskHQ/lips: proposals/lip-0045.md

| `MIN_CHAIN_NAME_LENGTH` | uint32 | 1 | The minimum length of a string specifying the name of a chain. |
| `MAX_CHAIN_NAME_LENGTH` | uint32 | 32 | The maximum length of a string specifying the name of a chain. |

We further use the utility function `getMainchainID` defined in [LIP 0037][lip-0037#getMainchainID] to obtain the chain ID of the mainchain.

> #### Empty Cross-chain Message

Reviewer: edited

### Genesis Block Processing

#### Genesis Assets Schema

```java
genesisInteroperabilityStoreSchema = {
    "type": "object",
    "required": [

>         "ownChainInfo",

Reviewer: Why not just calling it `ownChainAccount`?

### Genesis Block Processing

#### Genesis Assets Schema

```java
genesisInteroperabilityStoreSchema = {
    "type": "object",
    "required": [
        "ownChainName",
        "ownChainNonce",
        "chainInfos",
        "terminatedStateAccounts",
        "terminatedOutboxAccounts"
    ],
    "properties": {
        "ownChainName": {
            "dataType": "string",

>             "maxLength": MAX_CHAIN_NAME_LENGTH,

Reviewer: because if the `ownChainAccount` does not exist (can only happen on a sidechain), I can set this property to the default value of empty string. Having a minimum length was a bit awkward because then the empty string would not follow the schema

>             "maxLength": MAX_CHAIN_NAME_LENGTH,

Reviewer 2: makes sense

Here, the `...` notation, borrowed from [JavaScript ES6 data destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring), indicates that the corresponding schema should be inserted in place, and it is just used for notational convenience.

#### Genesis State Initialization

During the genesis state initialization stage of a genesis block `g`, the following steps are executed. If any step fails, the block is discarded and has no further effect.

Let `genesisBlockAssetBytes` be the `data` bytes included in the genesis block assets for the Interoperability module and let `interoperabilityAsset = decode(genesisInteroperabilityStoreSchema, genesisBlockAssetBytes)`. Let `chainInfos = interoperabilityAsset.chainInfos`.

##### Genesis Asset Verification

> * Within `chainInfos`, check that all entries have a different `chainID`.

Reviewer: I would also impose lexicographical order by `chainInfo.chainID` property.

> * For every `chainInfo` in `chainInfos`, properties `chainInfo.chainData`, `chainInfo.channelData` and `chainInfo.chainValidators` must be non-empty.

Reviewer: Here maybe we need to be a bit more precise. In my understanding, these properties are objects, so that 'empty' means that all properties are set to their default values. So I am wondering if this check is not actually redundant: for instance, the default value of the chain data would be invalid against the schema validation.

Reviewer: I split the properties into several arrays, to avoid this problem

> * For every `chainInfo` in `chainInfos` with non-empty `chainInfo.terminatedOutbox`, the field `chainInfo.terminatedState` must also be non-empty.

Reviewer: I think that also here this is not correct. What does 'empty' mean precisely in this context? I think you mean 'set to empty bytes', but this is true only if we have intermediate serialization, which is not what we do in the schema. The most elegant solution would be to have non-required `terminatedState` and `terminatedOutbox`, but this is not allowed by Lisk codec. Maybe we need intermediate serialization? Or maybe we can get away with 'empty', but we need to be very precise here

Reviewer: I split into several properties to avoid this problem

> * For every `chainInfo` in `chainInfos` with `chainInfo.terminatedState.initialized == False`, the field `chainInfo.terminatedOutbox` must be empty, `chainInfo.terminatedState.stateRoot` must be equal to `EMPTY_HASH` and `chainInfo.terminatedState.mainchainStateRoot` must not be equal to `EMPTY_HASH`. Conversely, if `chainInfo.terminatedState.initialized == True`, then `chainInfo.terminatedState.stateRoot` must not be equal to `EMPTY_HASH` and `chainInfo.terminatedState.mainchainStateRoot` must be equal to `EMPTY_HASH`. todo: must the terminated outbox be non-empty in this case?

Reviewer: There is a left over 'todo' comment

    * all `chainData.name` are unique.
    * all `chainData.name` have character set `a-z0-9!@$&_.`.
    * property `chainData.status` is in set `{CHAIN_STATUS_REGISTERED, CHAIN_STATUS_ACTIVE, CHAIN_STATUS_TERMINATED}`.
* For each entry `chainInfo` in `chainInfos`, let `activeValidators = chainInfo.chainValidators.activeValidators` and let `certificateThreshold = chainInfo.chainValidators.certificateThreshold`, then check:
    * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements.
    * `activeValidators` must be ordered lexicographically by `blsKey` property.
    * All `blsKey` properties must be pairwise distinct.  
    * Let `totalWeight` be the sum of the `bftWeight` property of every element in `activeValidators`. Then `totalWeight` has to be less than or equal to `MAX_UINT64`.
    * Check that `totalWeight//3 + 1 <= certificateThreshold <= totalWeight`, where `//` indicates integer division.
    * Check that the corresponding `validatorsHash` stored in `chainInfo.chainData.lastCertificate.validatorsHash` matches with the value computed from `activeValidators` and `certificateThreshold`.

> * On a sidechain, if `chainInfos` has an entry for another sidechain (`chainID != getMainchainID()`), then there must also exist an entry for the mainchain (`chainID == getMainchainID`) and `interoperabilityAsset.ownChainInfo` must be non-empty.

Reviewer: ```suggestion
* On a sidechain, if `chainInfos` has an entry for another sidechain (`chainID != getMainchainID()`), then there must also exist an entry for the mainchain (`chainID == getMainchainID()`) and `interoperabilityAsset.ownChainInfo` must be non-empty.
```

> * On a sidechain, if `chainInfos` has an entry for another sidechain (`chainID != getMainchainID()`), then there must also exist an entry for the mainchain (`chainID == getMainchainID`) and `interoperabilityAsset.ownChainInfo` must be non-empty.

Reviewer: I think we should still define the rules for the chainInfo.chainData entry for the mainchain, i.e. the fact that chainID and name must match.

Reviewer: This is now done

* For each entry `chainInfo` in `chainInfos`, check that `chainInfo.chainData.lastCertificate.timestamp < g.header.timestamp`.
* For each entry `chainInfo` in `chainInfos`, check that `chainInfo.chainId[0] == interoperabilityAsset.ownChainInfo.chainID[0]` and `chainInfo.chainId != interoperabilityAsset.ownChainInfo.chainID`.
* On the mainchain, check that `interoperabilityAsset.ownChainInfo.name == CHAIN_NAME_MAINCHAIN` and `interoperabilityAsset.ownChainInfo.chainID == getMainchainID()`.
* For each entry `chainInfo` in `chainInfos`, either `chainInfo.channelData.messageFeeTokenID == Token.getTokenIDLSK()` (corresponding to the LSK token), or `Token.getChainID(chainInfo.channelData.messageFeeTokenID)` must be equal to `chainInfo.chainID` or `interoperabilityAsset.ownChainInfo.chainID` (the message fee token must be a native token of one of the chains).

> ##### Genesis State Processing

Reviewer: Shouldn't this logic be run in the genesis state finalization? I am not sure, as maybe it would be better to completely initialize the internal state already here.

> * For every `chainInfo` in `chainInfos` with empty `chainInfo.terminatedState` add an entry to the outbox root substore with key `chainInfo.chainID` and value `chainInfo.channelData.outbox.root`. todo: Can `chainInfo` contain no outbox root after verification?

Reviewer: remove todo comment

> * For every `chainInfo` in `chainInfos` with empty `chainInfo.terminatedState` add an entry to the outbox root substore with key `chainInfo.chainID` and value `chainInfo.channelData.outbox.root`. todo: Can `chainInfo` contain no outbox root after verification?

Reviewer: We can be more precise here and use the `encode` notation we always use when we set something in the state. Also, we should be precise about definition of empty `chainInfo.terminatedState` (already mentioned above)

* Initialize own chain substore with data from `interoperabilityAsset.ownChainInfo`.
* For every entry `chainInfo` in `chainInfos` add the following substore entries with the key `chainInfo.chainID`:
    * with the value `chainInfo.chainData` to the chain data substore;
    * with the value `chainInfo.channelData` to the channel data substore;
    * with the value `chainInfo.chainValidators` to the chain validators substore;
    * with the value `chainInfo.terminatedState` to the terminated state substore, if `chainInfo.terminatedState` is not empty;
    * with the value `chainInfo.terminatedOutbox` to the terminated outbox substore, if `chainInfo.terminatedOutbox` is not empty.

> * On the mainchain, for every `chainInfo` in `chainInfos` add an entry to the registered names substore with key `chainInfo.chainData.name` and value `chainInfo.chainID`.

Reviewer: On the mainchain also add entry for the mainchain itself

##### Genesis Asset Verification

The genesis asset verification follows different rules on the mainchain or on a sidechain.

On the mainchain:

* `ownChainName == CHAIN_NAME_MAINCHAIN`.
* If `chainInfos` is non-empty, then `ownChainNonce > 0`. Conversely, if `chainInfos` is empty, then `ownChainNonce == 0`.
* Each entry `chainInfo` in `chainInfos` has a unique `chainInfo.chainID` and is ordered lexicographically by `chainInfo.chainID`. Furthermore for each entry it holds:
  * `chainInfo.chainID != getMainchainID()`;
  * `chainInfo.chainId[0] == getMainchainID()[0]`.
* For each entry `chainInfo` in `chainInfos`, let `chainData = chainInfo.chainData`, then check:
  * `chainData.lastCertificate.timestamp < g.header.timestamp`;

>   * all `chainData.name` are unique;

Reviewer: edited in a similar way

>   * all `chainData.name` are from the character set `a-z0-9!@$&_.` and length between `MIN_CHAIN_NAME_LENGTH` and `MAX_CHAIN_NAME_LENGTH`;

Reviewer: I added the check using the new API `validateObjectSchema`

>   * all `chainData.name` are from the character set `a-z0-9!@$&_.` and length between `MIN_CHAIN_NAME_LENGTH` and `MAX_CHAIN_NAME_LENGTH`;

Reviewer: and implemented your suggestion

  * property `chainData.status` is in set `{CHAIN_STATUS_REGISTERED, CHAIN_STATUS_ACTIVE, CHAIN_STATUS_TERMINATED}`.
* For each entry `chainInfo` in `chainInfos`, let `channelData = chainInfo.channelData`, then check:
  * `channelData.messageFeeTokenID == Token.getTokenIDLSK()`.
* For each entry `chainInfo` in `chainInfos`, let `activeValidators = chainInfo.chainValidators.activeValidators` and let `certificateThreshold = chainInfo.chainValidators.certificateThreshold`, then check:
  * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements;
  * `activeValidators` must be ordered lexicographically by `blsKey` property;
  * all `blsKey` properties must be pairwise distinct;  

>   * let `totalWeight` be the sum of the `bftWeight` property of every element in `activeValidators`. Then `totalWeight` has to be less than or equal to `MAX_UINT64`;

Reviewer 2: We must also check `bftWeight > 0`.

Reviewer: added

Reviewer 3: I would write " be the sum of the `bftWeight` property of all elements in `activeValidators`"

Reviewer 3: Same for sidechains.

* Each entry `outboxAccount` in `terminatedOutboxAccounts` has a unique `outboxAccount.chainID` and is ordered lexicographically by `outboxAccount.chainID`. Furthermore, an entry `outboxAccount` in `terminatedOutboxAccounts` must have a corresponding entry (i.e., with `chainID == outboxAccount.chainID`) in `terminatedStateAccounts`. Notice that the opposite is not necessarily true, so that there could be an entry in `terminatedStateAccounts` without a corresponding entry in `terminatedOutboxAccounts`.

On a sidechain, the Interoperability state can only contain the chain account for the mainchain (if the mainchain registration was done) or be empty. Hence, `chainInfos` is either empty or it contains exactly one entry for the mainchain.

If `chainInfos` is empty, then check that: 

* `ownChainName` is the empty string;
* `ownChainNonce == 0`;
* `terminatedStateAccounts` is empty;
* `terminatedOutboxAccounts` is empty.

If `chainInfos` is not empty, then check that:

* `ownChainName` is from the character set `a-z0-9!@$&_.`, has length between `MIN_CHAIN_NAME_LENGTH` and `MAX_CHAIN_NAME_LENGTH`, and `ownChainName != CHAIN_NAME_MAINCHAIN`;
* `ownChainNonce > 0`;

> * `chainInfos` contains exactly one entry with:

Reviewer 2: I would apply the same validation rules for `chainValidators` as on the mainchain.

Reviewer: added

  * `chainID == getMainchainID()`;
  * `chainData.name ==  CHAIN_NAME_MAINCHAIN`, `chainData.status !=  CHAIN_STATUS_TERMINATED`, and `chainData.lastCertificate.timestamp < g.header.timestamp`;

>   * `channelData.messageFeeTokenID == Token.getTokenIDLSK()`.

Reviewer 2: Also here, we should enforce `minReturnFeePerByte == MIN_RETURN_FEE_PER_BYTE_LSK`.

Reviewer: added

> * Each entry `stateAccount` in `terminatedStateAccounts` has a unique `stateAccount.chainID` and is ordered lexicographically by `stateAccount.chainID`. Furthermore for each entry it holds `stateAccount.chainID != getMainchainID()` and `stateAccount.chainID != ownChainAccount.chainID`.

Reviewer 2: ```suggestion
* Each entry `stateAccount` in `terminatedStateAccounts` has a unique `stateAccount.chainID` and `terminatedStateAccounts` is ordered lexicographically by `stateAccount.chainID`. Furthermore for each entry it holds `stateAccount.chainID != getMainchainID()` and `stateAccount.chainID != ownChainAccount.chainID`.
```

Here, the `...` notation, borrowed from [JavaScript ES6 data destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring), indicates that the corresponding schema should be inserted in place, and it is just used for notational convenience.

#### Genesis State Initialization

During the genesis state initialization stage of a genesis block `g`, the following steps are executed. If any step fails, the block is discarded and has no further effect.

> Let `genesisBlockAssetBytes` be the `data` bytes included in the genesis block assets for the Interoperability module and let `interoperabilityAsset = decode(genesisInteroperabilityStoreSchema, genesisBlockAssetBytes)`. Let `ownChainName = interoperabilityAsset.ownChainName`, `ownChainNonce = interoperabilityAsset.ownChainNonce`, `chainInfos = interoperabilityAsset.chainInfos`, `terminatedStateAccounts = interoperabilityAsset.terminatedStateAccounts`, and `terminatedOutboxAccounts = interoperabilityAsset.terminatedOutboxAccounts`.

Reviewer: I would not change this

###### Sidechain

On a sidechain, the Interoperability state can only contain the chain account for the mainchain (if the mainchain registration was done) or be empty. Hence, `chainInfos` is either empty or it contains exactly one entry for the mainchain.

If `chainInfos` is empty, then check that: 

* `ownChainName` is the empty string;
* `ownChainNonce == 0`;
* `terminatedStateAccounts` is empty;
* `terminatedOutboxAccounts` is empty.

If `chainInfos` is not empty, then check that:

* `ownChainName` is from the character set `a-z0-9!@$&_.`, has length between `MIN_CHAIN_NAME_LENGTH` and `MAX_CHAIN_NAME_LENGTH`, and `ownChainName != CHAIN_NAME_MAINCHAIN`;
* `ownChainNonce > 0`;

> * `chainInfos` contains exactly one entry `mainchainInfo` with:

Reviewer: In a sense it is always explicit in the sense that they cannot be created with the current protocol.

  * `mainchainInfo.channelData.messageFeeTokenID == Token.getTokenIDLSK()`;
  * `mainchainInfo.channelData.minReturnFeePerByte == MIN_RETURN_FEE_PER_BYTE_LSK`.
* Let `activeValidators = mainchainInfo.chainValidators.activeValidators` and let `certificateThreshold = mainchainInfo.chainValidators.certificateThreshold`, then check:
  * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements;
  * `activeValidators` must be ordered lexicographically by `blsKey` property;
  * all `blsKey` properties must be pairwise distinct;
  * for each `validator` in `activeValidators`, `validator.bftWeight > 0` must hold;
  * let `totalWeight` be the sum of the `bftWeight` property of every element in `activeValidators`. Then `totalWeight` has to be less than or equal to `MAX_UINT64`;
  * check that `totalWeight//3 + 1 <= certificateThreshold <= totalWeight`, where `//` indicates integer division;
  * check that the corresponding `validatorsHash` stored in `mainchainInfo.chainData.lastCertificate.validatorsHash` matches with the value computed from `activeValidators` and `certificateThreshold`.

> * Each entry `stateAccount` in `terminatedStateAccounts` has a unique `stateAccount.chainID` and `terminatedStateAccounts` is ordered lexicographically by `stateAccount.chainID`. Furthermore for each entry it holds `stateAccount.chainID != getMainchainID()` and `stateAccount.chainID != ownChainAccount.chainID`.

Reviewer: yes. As I mentioned, I would handle that separately (it is not clear to me that this can happen).

  * `chainInfo.chainID != getMainchainID()`;
  * `chainInfo.chainId[0] == getMainchainID()[0]`.
* For each entry `chainInfo` in `chainInfos`, let `chainData = chainInfo.chainData`. The entries `chainData.name` must be pairwise distinct. Furthermore for each entry it holds:
  * `chainData.lastCertificate.timestamp < g.header.timestamp`;
  * `chainData.name` only uses the character set `a-z0-9!@$&_.`;
  * property `chainData.status` is in set `{CHAIN_STATUS_REGISTERED, CHAIN_STATUS_ACTIVE, CHAIN_STATUS_TERMINATED}`.
* For each entry `chainInfo` in `chainInfos`, let `channelData = chainInfo.channelData`, then check:
  * `channelData.messageFeeTokenID == Token.getTokenIDLSK()`;
  * `channelData.minReturnFeePerByte == MIN_RETURN_FEE_PER_BYTE_LSK`.
* For each entry `chainInfo` in `chainInfos`, let `activeValidators = chainInfo.chainValidators.activeValidators` and let `certificateThreshold = chainInfo.chainValidators.certificateThreshold`, then check:
  * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements;
  * `activeValidators` must be ordered lexicographically by `blsKey` property;
  * all `blsKey` properties must be pairwise distinct;

>   * for each `validator` in `activeValidators`, check that `validator.bftWeight > 0`;

Reviewer 2: ```suggestion
  * for each `validator` in `activeValidators`, `validator.bftWeight > 0` must hold;
```

Reviewer 2: Otherwise, it's "then check: ... check that ..."

* `ownChainName` is from the character set `a-z0-9!@$&_.`, has length between `MIN_CHAIN_NAME_LENGTH` and `MAX_CHAIN_NAME_LENGTH`, and `ownChainName != CHAIN_NAME_MAINCHAIN`;
* `ownChainNonce > 0`;
* `chainInfos` contains exactly one entry `mainchainInfo` with:
  * `mainchainInfo.chainID == getMainchainID()`;
  * `mainchainInfo.chainData.name == CHAIN_NAME_MAINCHAIN`, `chainData.status != CHAIN_STATUS_TERMINATED`, and `chainData.lastCertificate.timestamp < g.header.timestamp`;
  * `mainchainInfo.channelData.messageFeeTokenID == Token.getTokenIDLSK()`;
  * `mainchainInfo.channelData.minReturnFeePerByte == MIN_RETURN_FEE_PER_BYTE_LSK`.
* Let `activeValidators = mainchainInfo.chainValidators.activeValidators` and let `certificateThreshold = mainchainInfo.chainValidators.certificateThreshold`, then check:
  * `activeValidators` must have at least 1 element and at most `MAX_NUM_VALIDATORS` elements;
  * `activeValidators` must be ordered lexicographically by `blsKey` property;
  * all `blsKey` properties must be pairwise distinct;

>   * for each `validator` in `activeValidators`, check that `validator.bftWeight > 0`;

Reviewer 2: ```suggestion
  * for each `validator` in `activeValidators`, `validator.bftWeight > 0` must hold;
```
